mender-auth/client/client.cpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-auth/client/client.hpp>
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/conf/paths.hpp>
+#include <common/crypto.hpp>
+#include <common/json.hpp>
+#include <common/error.hpp>
+#include <common/log.hpp>
+#include <common/path.hpp>
+#include <common/expected.hpp>
+#include <common/identity_parser.hpp>
+
+using namespace std;
+namespace error = mender::common::error;
+namespace common = mender::common;
+namespace conf = mender::common::conf;
+
+
+namespace identity_parser = mender::common::identity_parser;
+namespace key_value_parser = mender::common::key_value_parser;
+namespace config_path = mender::common::conf::paths;
+namespace path = mender::common::path;
+namespace mlog = mender::common::log;
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace crypto = mender::common::crypto;
+
+
+const string request_uri = "/api/devices/v1/authentication/auth_requests";
+
+
+const AuthClientErrorCategoryClass AuthClientErrorCategory;
+
+const char *AuthClientErrorCategoryClass::name() const noexcept {
+	return "AuthClientErrorCategory";
+}
+
+string AuthClientErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case SetupError:
+		return "Error during setup";
+	case RequestError:
+		return "HTTP client request error";
+	case ResponseError:
+		return "HTTP client response error";
+	case APIError:
+		return "API error";
+	case UnauthorizedError:
+		return "Unauthorized error";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(AuthClientErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, AuthClientErrorCategory), msg);
+}
+
+error::Error MakeHTTPResponseError(
```

Comment:
`inline`

Reply:
It's likely to be a cheaper operation. So just a good practice, I'd say.

Reply:
I would generally argue against using `inline`. Space is likely to be a much more important factor than running time for low level devices. So we don't want to inline things, at least not forcibly. Better to control this indirectly with `-O` options (which will inline appropriate things even if you didn't ask for it, see `-finline-small-functions`).

Reply:
The optimizer will also shrink the code so that the size impact will be negligible anyway. But :+1: